src/osd/PG.cc

```diff
@@ -7403,9 +7403,10 @@ boost::statechart::result PG::RecoveryState::Active::react(const MLogRec& logevt
     pg->peer_missing[logevt.from],
     logevt.from,
     context< RecoveryMachine >().get_recovery_ctx());
-  if (pg->is_peered() &&
-      got_missing)
-    pg->queue_recovery();
+  if (got_missing) {
+    post_event(DoRecovery());
```

Comment:
Seems we still needs some sanity checking here, there are multiple substates of **Active**(e.g., **WaitRemoteRecoveryReserved**) can not handle **DoRecovery()** event properly...

Reply:
Yes, see commit 64047e1

This issue came up in testing pull request ceph/ceph#19850 which should merge soon.